text suggestions (#4)
safe-research/permissionless-create2: doc/eip.md

---

> title: Permissionless CREATE2 Factory via EIP-7702

Reviewer: ```suggestion
title: Universal CREATE2 Factory
```

## Motivation

Ensuring that contracts share the same address and code on multiple chains is a hard problem. It is typically done by having a known CREATE2 factory contract at a specific address that can further deterministically deploy new contracts using the `CREATE2 (0xf5)` opcode.

However, there is a bootstrapping problem: how do you get a CREATE2 factory contract with a specific address and code? 

### Existing Solutions

There are currently three main approaches to this problem:

#### 1. Nick's method

Use Nick's method to randomly generate a signature for a transaction **without** [EIP-155](./eip-155.md) replay protection that deploys the CREATE2 factory. Nick's method ensures that there is no known private key for an account that deploys the CREATE2 factory, meaning that the resulting contract will have a deterministic address and code on all chains. This strategy is used by [Arachnid/deterministic-deployment-proxy](https://github.com/Arachnid/deterministic-deployment-proxy), one of the most widely used CREATE2 factory contracts.

**Downsides**:

- It does not work on chains that only accept EIP-155 replay-protected transactions.

> - It is sensitive to changes in gas parameters on the target chain since the gas price in the deployment transaction is sealed, and a new one cannot be signed without a private key.

Reviewer: ```suggestion
- It is sensitive to changes in gas parameters on the target chain since the gas price and limit in the deployment transaction is sealed, and a new one cannot be signed without a private key.
```

**Downsides**:

- It is not standardized nor adopted by all chains.
- It is permissioned as a chain can choose not to include a specific CREATE2 factory contract preinstalled.
- Attempts to standardize this with [RIP-7740](https://github.com/ethereum/RIPs/blob/master/RIPS/rip-7740.md) have not been successful.

### Proposal: Using type-4 transactions

This ERC proposes a permissionless alternative fourth mechanism to the existing ones described above with none of their downsides. Additionally, it standardizes a set of deployment parameters for a **universal** CREATE2 factory deployment. This ensures a common CREATE2 factory for the community instead of multiple competing copies with slightly different codes at different addresses. This single CREATE2 factory copy can bootstrap additional deterministic deployment infrastructure (such as the comprehensive CreateX universal contract deployer).

**Benefits**

- Universally applicable: It can be executed on any chain by any user and guarantees a reliable determination of smart contract deployments on any chain.
- Fault resistant: The method is secure against gas timeouts and other errors.

> - Permissionless: The method can be used by anyone on any EVM-compatible chain.

Reviewer: ```suggestion
- Permissionless: The universal CREATE2 factory contract can be deployed by anyone on any EVM-compatible chain.
```

## Abstract

> This ERC defines a permissionless and deterministic deployment mechanism for a universal CREATE2 factory contract with a fixed cross-chain address (0xC0DE207acb0888c5409E51F27390Dad75e4ECbe7) and known bytecode. By leveraging EIP-7702 for stateless delegation and EIP-1014's CREATE2 (0xf5) opcode, this standard enables trustless, reproducible contract deployments to deterministic addresses across all EVM-compatible chains—without relying on preinstalls, secret keys, or chain-specific infrastructure.

Reviewer: ```suggestion
This ERC defines a permissionless and deterministic deployment mechanism for a universal CREATE2 factory contract with a fixed cross-chain address (0xC0DE207acb0888c5409E51F27390Dad75e4ECbe7) and known bytecode. By leveraging EIP-7702 for stateless delegation and EIP-1014's `CREATE2 (0xf5)` opcode, this standard enables trustless, reproducible contract deployments to deterministic addresses across all EVM-compatible chains—without relying on preinstalls, secret keys, or chain-specific infrastructure.
```

## Motivation

Ensuring that contracts share the same address and code on multiple chains is a hard problem. It is typically done by having a known CREATE2 factory contract at a specific address that can further deterministically deploy new contracts using the `CREATE2 (0xf5)` opcode.

However, there is a bootstrapping problem: how do you get a CREATE2 factory contract with a specific address and code? 

### Existing Solutions

There are currently three main approaches to this problem:

#### 1. Nick's method

Use Nick's method to randomly generate a signature for a transaction **without** [EIP-155](./eip-155.md) replay protection that deploys the CREATE2 factory. Nick's method ensures that there is no known private key for an account that deploys the CREATE2 factory, meaning that the resulting contract will have a deterministic address and code on all chains. This strategy is used by [Arachnid/deterministic-deployment-proxy](https://github.com/Arachnid/deterministic-deployment-proxy), one of the most widely used CREATE2 factory contracts.

**Downsides**:

   - It does not work on chains that only accept EIP-155 replay-protected transactions.

>    - It is sensitive to changes in gas parameters on the target chain since the gas price in the deployment transaction is sealed, and a new one cannot be signed without a private key.

Reviewer: nit: I wouldn't indent these lines. I don't think it makes a difference for Markdown rendering.

   - Reverts due to alternative gas schedules make the CREATE2 factory no longer deployable.

#### 2. Secret private key

Keep a carefully guarded secret key and use it to sign transactions to deploy CREATE2 factory contracts. The resulting contract will have a deterministic address and code on all chains where the first transaction of the deployer account is a CREATE2 factory deployment, which can be verified post-deployment to ensure trustlessness. This is the strategy used by [safe-global/safe-singleton-factory](https://github.com/safe-global/safe-singleton-factory) and [pcaversaccio/createx](https://github.com/pcaversaccio/createx). 

**Downsides**:

   - It is permissioned: the party that holds the secret key has the ultimate say on which chains will get the CREATE2 factory deployments.

>    This requires carefully guarding a secret key; if it is exposed or lost, deployments are no longer guaranteed on new chains.

Reviewer: ```suggestion
   - This requires carefully guarding a secret key; if it is exposed or lost, deployments are no longer guaranteed on new chains.
```

   - If the first transaction is not a successful CREATE2 factory deployment, then it is no longer possible to have a CREATE2 factory at the common address; this can happen by human error, for example.

#### 3. CREATE2 deployment factories

> Have popular CREATE2 deployment factories deployed on new chains by default. This is, for example, what OP Stack does as part of their [preinstalls](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/libraries/Preinstalls.sol), including the CREATE2 factory contracts mentioned above. This ensures that the CREATE2 factory contracts have known addresses and codes.

Reviewer: Not related to your change, but I noticed that this should probably be a permalink:

```suggestion
Have popular CREATE2 deployment factories deployed on new chains by default. This is, for example, what OP Stack does as part of their [preinstalls](https://github.com/ethereum-optimism/optimism/blob/12c5398a1725a2aafc3e7abb0711cf761a2b20b1/packages/contracts-bedrock/src/libraries/Preinstalls.sol), including the CREATE2 factory contracts mentioned above. This ensures that the CREATE2 factory contracts have known addresses and codes.
```

**Downsides**:

   - It is not standardized nor adopted by all chains.
   - It is permissioned as a chain can choose not to include a specific CREATE2 factory contract preinstalled.
   - Attempts to standardize this with [RIP-7740](https://github.com/ethereum/RIPs/blob/master/RIPS/rip-7740.md) have not been successful.

> This ERC proposes a permissionless alternative fourth mechanism to the existing ones described above with none of their downsides. Additionally, it standardizes a set of deployment parameters for a **unique** CREATE2 factory deployment. This ensures a common CREATE2 factory for the community instead of multiple competing copies with slightly different codes at different addresses. This single CREATE2 factory copy can bootstrap additional deterministic deployment infrastructure (such as the comprehensive CreateX universal contract deployer).

Reviewer: I feel like this sentence and the section below have some uneccessary duplication. Also not a fan that this is technically in the '3. CREATE2 deployment factories' section despite not belonging there. Maybe it can be merged with the section below?

> ### Proposal: Using type-4 transactions

Reviewer: ```suggestion
### Proposal: Using EIP-7702 Type `0x4` transactions
```

From some other EIPs/ERCs I've read, this is how they are written.

This ERC proposes a fourth solution: A publicly available deployer address uses a type-4 delegation from EIP-7702 to execute the CREATE2 call from a standardized bootstrap contract.

**Benefits**

  Universally applicable: It can be executed on any chain by any user and guarantees a reliable determination of Smart Contract deployments on any chain.
  - Fault resistant: The method is secure against gas timeouts and other errors.

> 

Reviewer: permissionless
